Implement WebTransport incoming unidirectional and bidirectional streams

AnkshitJain opened the pull request:
#### c84ec7e6f2079d76b34158a6085979defb7ceaa9
<pre>
Implement WebTransport incoming unidirectional and bidirectional streams
<a href="https://bugs.webkit.org/show_bug.cgi?id=284501">https://bugs.webkit.org/show_bug.cgi?id=284501</a>
<a href="https://rdar.apple.com/136262852">rdar://136262852</a>

Reviewed by Alex Christensen.

Incoming WebTransport streams are hooked up to the WebTransport session so data can be sent and received on them.
Two tests are added to validate incoming unidirectional and bidirectional streams.
* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::WebTransport::receiveIncomingUnidirectionalStream):
(WebCore::WebTransport::receiveBidirectionalStream):
* Source/WebCore/Modules/webtransport/WebTransport.h:
(WebCore::WebTransport::scriptExecutionContext const):
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStream.idl:
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamSource.cpp:
(WebCore::WebTransportBidirectionalStreamSource::receiveIncomingStream):
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamSource.h:
* Source/WebCore/Modules/webtransport/WebTransportReceiveStream.idl:
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.cpp:
(WebCore::WebTransportReceiveStreamSource::receiveIncomingStream):
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.h:
* Source/WebCore/Modules/webtransport/WebTransportSessionClient.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::receiveIncomingUnidirectionalStream):
(WebKit::NetworkTransportSession::receiveBidirectionalStream):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::NetworkTransportSession::setupConnectionHandler):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm:
(WebKit::NetworkTransportStream::receiveLoop):
* Source/WebKit/WebProcess/Network/WebTransportReceiveStreamSource.cpp:
(WebKit::WebTransportReceiveStreamSource::WebTransportReceiveStreamSource):
(WebKit::WebTransportReceiveStreamSource::~WebTransportReceiveStreamSource):
* Source/WebKit/WebProcess/Network/WebTransportReceiveStreamSource.h:
(WebKit::WebTransportReceiveStreamSource::create):
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::receiveIncomingUnidirectionalStream):
(WebKit::WebTransportSession::receiveBidirectionalStream):
(WebKit::WebTransportSession::createBidirectionalStream):
* Tools/TestWebKitAPI/NetworkConnection.h:
(TestWebKitAPI::ConnectionGroup::ConnectionGroup):
* Tools/TestWebKitAPI/NetworkConnection.mm:
(TestWebKitAPI::ConnectionGroup::createWebTransportConnection const):
(TestWebKitAPI::ConnectionGroup::terminate):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, DISABLE_ClientBidirectional)):
(TestWebKitAPI::TEST(WebTransport, DISABLE_Datagram)):
(TestWebKitAPI::TEST(WebTransport, DISABLE_Unidirectional)):
(TestWebKitAPI::TEST(WebTransport, DISABLE_ServerBidirectional)):
(TestWebKitAPI::TEST(WebTransport, DISABLED_ClientBidirectional)): Deleted.
(TestWebKitAPI::TEST(WebTransport, DISABLED_Datagram)): Deleted.
* Tools/TestWebKitAPI/WebTransportServer.h:
* Tools/TestWebKitAPI/WebTransportServer.mm:
(TestWebKitAPI::WebTransportServer::Data::create):
(TestWebKitAPI::WebTransportServer::Data::Data):
(TestWebKitAPI::WebTransportServer::WebTransportServer):
(TestWebKitAPI::WebTransportServer::port const):

Canonical link: <a href="https://commits.webkit.org/287777@main">https://commits.webkit.org/287777@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1cf51c505e5baecd82ce17e26aa2f1ff751c47a0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80798 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/320 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34734 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85325 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31782 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82909 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/337 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8119 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/63092 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/20878 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83867 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73542 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43395 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27702 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30239 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28235 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86757 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8025 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/5664 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/71391 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8202 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69378 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70631 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14657 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/13594 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12531 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7987 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/13508 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7826 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11345 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9632 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Tools/TestWebKitAPI/WebTransportServer.mm:
```diff
@@ -72,36 +71,25 @@
+    nw_listener_set_new_connection_group_handler(listener.get(), [
+        data = m_data
+    ] (nw_connection_group_t incomingConnectionGroup)
+    {
```
This should be on the previous line.  In fact, could you put the last 4 lines on 1 line?

nw_listener_set_new_connection_group_handler(listener.get(), [data = m_data] (nw_connection_group_t incomingConnectionGroup) {

Commit:
Implement WebTransport incoming unidirectional and bidirectional streams
https://bugs.webkit.org/show_bug.cgi?id=284501
rdar://136262852

Reviewed by Alex Christensen.

Incoming WebTransport streams are hooked up to the WebTransport session so data can be sent and received on them.
Two tests are added to validate incoming unidirectional and bidirectional streams.
* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::WebTransport::receiveIncomingUnidirectionalStream):
(WebCore::WebTransport::receiveBidirectionalStream):
* Source/WebCore/Modules/webtransport/WebTransport.h:
(WebCore::WebTransport::scriptExecutionContext const):
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStream.idl:
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamSource.cpp:
(WebCore::WebTransportBidirectionalStreamSource::receiveIncomingStream):
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamSource.h:
* Source/WebCore/Modules/webtransport/WebTransportReceiveStream.idl:
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.cpp:
(WebCore::WebTransportReceiveStreamSource::receiveIncomingStream):
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.h:
* Source/WebCore/Modules/webtransport/WebTransportSessionClient.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::receiveIncomingUnidirectionalStream):
(WebKit::NetworkTransportSession::receiveBidirectionalStream):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::NetworkTransportSession::setupConnectionHandler):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm:
(WebKit::NetworkTransportStream::receiveLoop):
* Source/WebKit/WebProcess/Network/WebTransportReceiveStreamSource.cpp:
(WebKit::WebTransportReceiveStreamSource::WebTransportReceiveStreamSource):
(WebKit::WebTransportReceiveStreamSource::~WebTransportReceiveStreamSource):
* Source/WebKit/WebProcess/Network/WebTransportReceiveStreamSource.h:
(WebKit::WebTransportReceiveStreamSource::create):
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::receiveIncomingUnidirectionalStream):
(WebKit::WebTransportSession::receiveBidirectionalStream):
(WebKit::WebTransportSession::createBidirectionalStream):
* Tools/TestWebKitAPI/NetworkConnection.h:
(TestWebKitAPI::ConnectionGroup::ConnectionGroup):
* Tools/TestWebKitAPI/NetworkConnection.mm:
(TestWebKitAPI::ConnectionGroup::createWebTransportConnection const):
(TestWebKitAPI::ConnectionGroup::terminate):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, DISABLE_ClientBidirectional)):
(TestWebKitAPI::TEST(WebTransport, DISABLE_Datagram)):
(TestWebKitAPI::TEST(WebTransport, DISABLE_Unidirectional)):
(TestWebKitAPI::TEST(WebTransport, DISABLE_ServerBidirectional)):
(TestWebKitAPI::TEST(WebTransport, DISABLED_ClientBidirectional)): Deleted.
(TestWebKitAPI::TEST(WebTransport, DISABLED_Datagram)): Deleted.
* Tools/TestWebKitAPI/WebTransportServer.h:
* Tools/TestWebKitAPI/WebTransportServer.mm:
(TestWebKitAPI::WebTransportServer::Data::create):
(TestWebKitAPI::WebTransportServer::Data::Data):
(TestWebKitAPI::WebTransportServer::WebTransportServer):
(TestWebKitAPI::WebTransportServer::port const):

Canonical link: https://commits.webkit.org/287777@main
